source/docs/software/wpilib-tools/robot-simulation/device-sim.rst

```diff
@@ -0,0 +1,113 @@
+
+.. note:: This example will use the ``EncoderSim`` class as an example. Use of other simulation classes will be almost identical.
+
+Creating Simulation Device objects
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Simulation device object can be constructed in two ways:
+- a constructor that accepts the regular hardware object.
+- a constructor or factory method that accepts the port/index/channel number that the device is connected to. These would be the same number that was used to construct the regular hardware object. This is especially useful for :doc:`unit testing <unit-testing>`.
+
+.. tabs::
+   .. code-tab:: java
+
+      // create a real encoder object on DIO 2,3
+      Encoder encoder = new Encoder(2, 3);
+      // create a sim controller for the encoder
+      EncoderSim simEncoder = new EncoderSim(encoder);
+
+   .. code-tab:: cpp
+
+       // create a real encoder object on DIO 2,3
+       frc::Encoder encoder{2, 3};
+       // create a sim controller for the encoder
+       frc::sim::EncoderSim simEncoder{encoder};
+
+Reading and Writing Device Data
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Each simulation class has getter (``getXxx()``/``GetXxx()``) and setter (``setXxx(value)``/``SetXxx(value)``) functions for each field ``Xxx``. The getter functions will return the same as the getter of the regular device class.
+
+.. tabs::
+   .. code-tab:: java
+
+      simEncoder.setCount(100);
+      encoder.getCount(); // 100
+      simEncoder.getCount(); // 100
+
+   .. code-tab:: cpp
+
+      simEncoder.SetCount(100);
+      encoder.GetCount(); // 100
+      simEncoder.GetCount(); // 100
+
+Registering Callbacks
+^^^^^^^^^^^^^^^^^^^^^
+
+In addition to the getters and setters, each field also has a ``registerXxxCallback()`` function that registers a callback to be run whenever the field value changes and returns a ``CallbackStore`` object. In C++, save this object in the right scope - the callback will be canceled when the object goes out of scope and is destroyed. In Java, keep a reference to the object so it isn't garbage-collected (the callback will be canceled if it is); and call  ``close()`` to cancel the callback. The callbacks accept a string parameter of the name of the field and a ``HALValue`` object containing the new value. In C++, another ``param`` parameter can be used to pass in arbitrary data - in Java, this data should be captured in the lambda. Before retrieving values from a ``HALValue``, check the type of value contained. Possible types are ``HALValue.kBoolean``/``HAL_BOOL``, ``HALValue.kDouble``/``HAL_DOUBLE``, ``HALValue.kEnum``/``HAL_ENUM``, ``HALValue.kInt``/``HAL_INT``, ``HALValue.kLong``/``HAL_LONG``.
```

Comment:
Perhaps just split this up into two separate paragraphs for Java/C++. Reading this made my brain hurt.